File: r/minimax.R
#' Given a matrix of utilities for two players returns the minimax solution for the row player.
#' 
#' Matching Pennies:
#' minimax_2(matrix(c(1, -1, -1, 1), nrow = 2))$solution
minimax_2 <- function(utilities) {
  require(dplyr)
  require(ROI)
  require(ROI.plugin.glpk)
  require(ompr)
  require(ompr.roi)
  
  row_strats <- 1:dim(utilities)[1]
  col_strats <- 1:dim(utilities)[2]
  
  MIPModel() %>%
    add_variable(row_utility, type = "continuous") %>%
    add_variable(row_pr[i], i = row_strats, type = "continuous", lb = 0, ub = 1) %>%
    set_objective(row_utility, "max") %>%
    add_constraint(sum_expr(row_pr[i]*utilities[i, j], i = row_strats) >= row_utility,
                   j = col_strats) %>%
    add_constraint(sum_expr(row_pr[i], i = row_strats) == 1) %>%
    solve_model(with_ROI(solver = "glpk"))
}

#' Given a matrix of utilities for three players returns the minimax solution for the first player.
#' !!Probably not working!!
minimax_3 <- function(utilities) {
  require(ROI)
  require(ROI.plugin.glpk)
  require(ompr)
  require(ompr.roi)
  
  x_strats <- 1:dim(utilities)[1]
  y_strats <- 1:dim(utilities)[2]
  z_strats <- 1:dim(utilities)[3]
  
  MIPModel() %>%
    add_variable(x_u, type = "continuous") %>%
    add_variable(x_pr[i], i = x_strats, type = "continuous", lb = 0, ub = 1) %>%
    set_objective(x_u, "max") %>%
    add_constraint(sum_expr(x_pr[i]*utilities[i, j, k], i = x_strats) >= x_u,
                   j = y_strats, k = z_strats) %>%
    add_constraint(sum_expr(x_pr[i], i = x_strats) == 1) %>%
    solve_model(with_ROI(solver = "glpk"))
}

File: jl/minimax.jl
# Minimax function for n players. Assumes players can all play the same strats.
# Ben Ewing on 2018-12-03

# Setup ------------------------------------------------------------------------
# Libraries
using JuMP, Distributions, LinearAlgebra, GLPKMathProgInterface, Combinatorics

# Two player minimax -----------------------------------------------------------
function minimax_2(knaps)
    # Solver
    m = Model(solver = GLPKSolverMIP())
    # number of strategies
    nstrat = size(knaps)[1]
    # player utility
    @variable(m, utility)
    # probability of playing a given knapsack
    @variable(m, knap_pr[1:nstrat] >= 0)
    # Objective is to maximize utility
    @objective(m, Max, utility)
    # s.t. probability sums to 1
    @constraint(m, sum(knap_pr) == 1)
    # s.t. utility determined by knapsacks
    for j in 1:nstrat
        @constraint(m, sum{knap_pr[i]*knaps[i, j], i = 1:nstrat} >= utility)
    end
    status = solve(m)
    getvalue(knap_pr)
end

# Three player minimax ---------------------------------------------------------
function minimax_3(knaps)
    # Solver
    m = Model(solver = GLPKSolverMIP())
    # number of strategies, assume square in all directions
    nstrat = size(knaps)[1]
    # player utility
    @variable(m, utility)
    # probability of playing a given knapsack
    @variable(m, knap_pr[1:nstrat] >= 0)
    # Objective is to maximize utility
    @objective(m, Max, utility)
    # s.t. probability sums to 1
    @constraint(m, sum(knap_pr) == 1)
    # s.t. utility determined by knapsacks
    for k in 1:nstrat
        for j in 1:nstrat
            @constraint(m, sum{knap_pr[i]*knaps[i, j, k], i = 1:nstrat} >= utility)
        end
    end
    status = solve(m)
    getvalue(knap_pr)
end

# n player minimax -------------------------------------------------------------
function minimax(knaps)
    # Solver
    m = Model(solver = GLPKSolverMIP())
    # number of strategies
    nstrat = size(knaps)[1]
    n_player = ndims(knaps)
    # player utility
    @variable(m, utility)
    # probability of playing a given knapsack
    @variable(m, knap_pr[1:nstrat] >= 0)
    # Objective is to maximize utility
    @objective(m, Max, utility)
    # s.t. probability sums to 1
    @constraint(m, sum(knap_pr) == 1)
    # s.t. utility determined by knapsacks

    for j in multiset_permutations(repeat(collect(1:nstrat), n_player), n_player-1)
        @constraint(m, sum{knap_pr[i]*knaps[i, j...], i = 1:nstrat} >= utility)
    end

    status = solve(m)
    getvalue(knap_pr)
end

# Testing ----------------------------------------------------------------------
# Two player
# Matching Pennies utilities
# minimax_2(hcat([1,-1], [-1,1]))
# minimax(hcat([1,-1], [-1,1]))
# rock paper scissors
# minimax_2(hcat([0, 1, -1], [-1, 0, 1], [1, -1, 0]))
# minimax(hcat([0, 1, -1], [-1, 0, 1], [1, -1, 0]))
# Three player
# 1-2-3-4
# minimax_3(cat([1 2; 3 4], [-1 -2; -3 -4], dims = 3))
# minimax(cat([1 2; 3 4], [-1 -2; -3 -4], dims = 3))
